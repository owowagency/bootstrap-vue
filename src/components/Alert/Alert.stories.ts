import Alert from '.';
import {variants} from '@/composables/useVariant';

export default {
    title: 'Components/Alerts',
    component: Alert,
    argTypes: {
        variant: {
            control: {type: 'select'},
            options: variants,
        },
    },
};

const template = (args) => ({
    components: {Alert},
    setup: () => ({args}),
    template: '<Alert v-bind="args">I am an alert</Alert>',
});

export const Default = template.bind({});

const variantsTemplate = (args) => ({
    components: {Alert},
    setup: () => ({args, variants}),
    template: '<Alert v-for="v of variants" class=" mb-1" :variant="v" v-bind="args">A simple {{ v }} alert - <a href="#" class="alert-link">Check it out!</a></Alert>',
});

export const Variants = variantsTemplate.bind({});

export const AdditionalContent = (args) => ({
    components: {Alert},
    setup: () => ({args}),
    template: `
        <Alert v-bind="args">
            <h4 class="alert-heading">Well done!</h4>
            <p>Aww yeah, you successfully read this important alert message. This example text is going to run a bit longer so that you can see how spacing within an alert works with this kind of content.</p>
            <hr>
            <p class="mb-0">Whenever you need to, be sure to use margin utilities to keep things nice and tidy.</p>
        </Alert>
    `,
});
