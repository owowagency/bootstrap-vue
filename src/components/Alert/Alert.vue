<template>
    <div
        ref="alert"
        class="alert"
        :class="classes"
    >
        <!-- @slot Displays content inside the alert -->
        <slot />
    </div>
</template>

<script lang="ts">
import {computed, ref} from 'vue';
import useBootstrapEmits from '@/composables/useBootstrapEmits';
import useBootstrapInstance from '@/composables/useBootstrapInstance';
import useClasses from '@/composables/useClasses';
import useVariant from '@/composables/useVariant';
import {variantProps} from '@/composables/useVariant';

const alertEvents = ['close', 'closed'] as const;

type AlertEvent = typeof alertEvents[number];
</script>

<script lang="ts" setup>
const props = defineProps(variantProps);

const emit = defineEmits<{(event: AlertEvent): void}>();

const alert = ref<HTMLElement>();

useBootstrapEmits(
    alert,
    alertEvents,
    emit,
    'alert',
);

const {bsInstance: bsAlert} = useBootstrapInstance(
    'Alert',
    alert,
);

const {classes} = useClasses(computed(() => [
    useVariant(props.variant, 'alert-{0}').variantClass.value,
]));

defineExpose({bsAlert});
</script>

<docs>
```vue
<Alert variant="danger">
    Something went wrong!
</Alert>
```

## Imported Props

| Prop name | Description                  | Type                                             | Values | Default   |
| --------- | ---------------------------- | ------------------------------------------------ | ------ | --------- |
| variant   | The variant of the component | [Variant](../../composables/useVariant) (string) | -      | 'primary' |

<!-- TODO Auto-generate based on type? -->
## Events

| Name   | Args | Description                                                             |
| ------ | ---- | ----------------------------------------------------------------------- |
| close  |      | Fires immediately when the `close` instance method is called            |
| closed |      | Fired when the alert has been closed and CSS transitions have completed |
</docs>
