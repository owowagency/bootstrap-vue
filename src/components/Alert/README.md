# Alert

```vue
<Alert variant="danger">
    Something went wrong!
</Alert>
```

## Imported Props

| Prop name | Description                  | Type                                             | Values | Default   |
| --------- | ---------------------------- | ------------------------------------------------ | ------ | --------- |
| variant   | The variant of the component | [Variant](../../composables/useVariant) (string) | -      | 'primary' |

<!-- TODO Auto-generate based on type? -->

## Events

| Name   | Args | Description                                                             |
| ------ | ---- | ----------------------------------------------------------------------- |
| close  |      | Fires immediately when the `close` instance method is called            |
| closed |      | Fired when the alert has been closed and CSS transitions have completed |

## Slots

| Name    | Description                       | Bindings |
| ------- | --------------------------------- | -------- |
| default | Displays content inside the alert |          |
