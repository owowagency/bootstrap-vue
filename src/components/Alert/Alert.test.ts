import {describe, expect, it, vi} from 'vitest';
import Alert from '.';
import {shallowMount} from '@vue/test-utils';

vi.mock('bootstrap', () => ({
    Alert: {
        getOrCreateInstance: vi.fn(),
    },
    __v_isShallow: true,
}));


describe('template', () => {
    componentRenderTest(Alert);

    componentSlotRenderTest(Alert);

    componentWrapperClassTest(Alert, {variant: 'danger'}, 'alert-danger');
});

describe('onMounted', () => {
    it('sets bsAlert', async() => {
        const wrapper = await shallowMount(Alert);

        expect((await import('bootstrap')).Alert.getOrCreateInstance).toBeCalledWith(wrapper.vm.$refs.alert, {});

        // TODO: Unable to assert bsAlert value.
    });
});
