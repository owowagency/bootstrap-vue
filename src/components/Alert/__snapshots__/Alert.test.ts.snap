// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`template > renders default 1`] = `
"<div class=\\"alert alert-primary\\">
  <!-- @slot Displays content inside the alert -->
</div>"
`;
